### ISSUE TITLE
Repaint scheduling

### ISSUE THREAD

**Issue by emersion** [swaywm/wlroots#655]
The idea is to send a frame event to clients, give them time to render a frame, and then render + swap buffers.

See https://www.collabora.com/about-us/blog/2015/02/12/weston-repaint-scheduling/  
See https://github.com/swaywm/wlroots/issues/644#issuecomment-366547068

**Comment by SirCmpwn** [comment]
I feel like we already had an issue for this somewhere...

**Comment by emersion** [comment]
#617 is different but related.

**Comment by emersion** [comment]
```
15:25 <MrCooper> pq: I guess ideally, the repaint window should be adjusted dynamically in response to hitting or missing the target
15:27 <pq> MrCooper, yes, more preferably by looking at the composite task fence completion timestamps.
15:27 <MrCooper> right
```

**Comment by YaLTeR** [comment]
I'm looking into this. I added timeline logging based on the linked Weston patch. Here are my observations so far:
- The reported DRM presentation timestamp can be before and after a `CLOCK_MONOTONIC` measurement inside `page_flip_handler()`, depending on the monitor.
- The minimum time from the return of rootston's `output_render()` to a vblank (basically how far `output_render()` can be delayed) which doesn't drop a frame varies heavily depending on the size of the updating region. With `glxgears` and `weston-simple-egl` at the default small size I am able to delay `output_render()` all the way to 2 milliseconds between `output_render()` returning and vblank without dropped frames, but at 1920×1080 fullscreen that required time increases to 5 milliseconds with 4 milliseconds already causing dropped frames about half of the time.

With that said, with a naive implementation of repaint delay (delays `output_render()` similarly to the Weston post) and frame delay (moves the frame callback from the end of `output_render()` to the specified number of milliseconds after a vblank), and additionally blocking all idle frames between a vblank and an `output_render()`, I was able to achieve sub-frame latency on rootston for clients relying on the frame event (like vsynced XWayland clients):

![image](https://user-images.githubusercontent.com/1794388/61168557-ddbb4280-a504-11e9-9077-2c12e144f12b.png)

Here the frame delay is set to 1 millisecond and repaint delay is set to 2 milliseconds, so the client `[id:1]` has about a little more than 1 millisecond to render, which it successfully does.
